src/opencl/opencl.cpp

```diff
@@ -0,0 +1,762 @@
+                if (err != SLOTH_NO_OPENCL_COMPATIBLE_NVIDIA_GPUS)
+                    OCL_ERR_CHECK(err, instances);
+                break;
+            case AMD:
+                devices = getAllAMDGPUs(err);
+                if (err != SLOTH_NO_OPENCL_COMPATIBLE_AMD_GPUS)
+                    OCL_ERR_CHECK(err, instances);
+                break;
+            case INTEL:
+                devices = getAllIntelGPUs(err);
+                if (err != SLOTH_NO_OPENCL_COMPATIBLE_INTEL_GPUS)
+                    OCL_ERR_CHECK(err, instances);
+                break;
+            default:
+                // Should never happen since functions in opencl_utils.hpp only get
+                // GPUs from only Nvidia, AMD or Intel platforms
+                assert(false);
+                break;
+        }
+        err = CL_SUCCESS;
+
+        for (size_t j = 0; j < devices.size(); j++) {
+            EncodeOpenCLInstance* instance = instances->instances + k;
+            instance->platform = platform;
+
+            instance->device = devices[j];
+
+            instance->context = clCreateContext(NULL, 1, &instance->device, NULL, NULL, &err);
+            OCL_ERR_CHECK(err, instances);
+
+            // Combine the relevant headers and kernel files together into a single string
+            std::string header_str = "";
+            if (instance->platform == NVIDIA) {
+                std::string ptx_str = std::string(mod256_189_cu);
+                header_str = ptx_str + "\n" + std::string(nvidia_specific_cl);
+            }
+            else {
+                header_str = std::string(non_nvidia_cl);
+            }
+            std::string kernel_str = std::string(encode_cl);
+            OCL_ERR_CHECK(err, instances);
+
+            std::string merged_str = header_str + "\n" + kernel_str;
+            size_t kernel_size = merged_str.size();
+            const char* c_kernel_str = merged_str.c_str();
+
+            instance->program = clCreateProgramWithSource(instance->context, 1,
+                                                          (const char**)&c_kernel_str,
+                                                          (const size_t*)&kernel_size,
+                                                          &err);
+            OCL_ERR_CHECK(err, instances);
+
+            // Compile the OpenCL kernel
+            std::string options = "";
+            switch (instance->platform) {
+                case NVIDIA:
+                    err = clBuildProgram(instance->program, 1, &instance->device, options.data(), NULL, NULL);
+                    if (err != CL_SUCCESS) {
+                        printProgramBuildLog(err, instance->device, instance->program);
+                        return instances;
+                    }
+                    break;
+                case AMD:
+                    size_t amd_wavefront_size;
+                    clGetDeviceInfo(instance->device,
+                                    CL_DEVICE_WAVEFRONT_WIDTH_AMD,

```

Comment:
I've run the basic opencl test on my MacBook. It works. 
However, this line creates an error when I try to build the code. It says `src/opencl/opencl.cpp:321:37: error: use of undeclared identifier 'CL_DEVICE_WAVEFRONT_WIDTH_AMD'`